File: D/697-degree-of-an-array/main.go
package main

import "fmt"

func findShortestSubArray(nums []int) int {
	MC := map[int]int{}
	MS := map[int]int{}
	ME := map[int]int{}
	for i := 0; i < len(nums); i++ {
		b := nums[i]

		MC[b]++
		_, ok := MS[b]
		if !ok {
			MS[b] = i
			ME[b] = i
		} else {
			ME[b] = i
		}
	}
	max := -1
	for _, v := range MC {
		if v > max {
			max = v
		}
	}

	ret := len(nums)
	for k, v := range MC {
		if v == max {
			if ME[k]-MS[k] < ret {
				ret = ME[k] - MS[k]
			}
		}
	}
	return ret + 1
}

func main() {
	fmt.Println("a")
}

File: D/697-degree-of-an-array/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn find_shortest_sub_array(nums: Vec<i32>) -> i32 {
    let mut m: HashMap<i32, i32> = HashMap::new();
    let mut p: HashMap<i32, usize> = HashMap::new();
    let mut ret = nums.len() + 1;
    let mut maxcnt = 0;
    for i in 0..nums.len() {
      let cnt = m.entry(nums[i]).or_insert(0);
      *cnt += 1;
      if *cnt > maxcnt {
        maxcnt = *cnt
      }
      p.entry(nums[i]).or_insert(i);
    }
    for i in (0..nums.len()).rev() {
      if m.contains_key(&nums[i]) && *(m.get(&nums[i]).unwrap()) == maxcnt {
        m.remove(&nums[i]);
        if i - *(p.get(&nums[i]).unwrap()) < ret {
          ret = i - p.get(&nums[i]).unwrap();
        }
      }
    }
    ret as i32 + 1
  }
}

fn main() {
  println!("{}",Solution::find_shortest_sub_array(vec![0;5]);)
}
